odxtools/parameters/tablekeyparameter.py

```diff
@@ -37,37 +40,107 @@ def _resolve_odxlinks(self, odxlinks: OdxLinkDatabase) -> None:
         super()._resolve_odxlinks(odxlinks)
 
         if self.table_ref:
-            self._table = odxlinks.resolve(self.table_ref)
+            if TYPE_CHECKING:
+                self._table = odxlinks.resolve(self.table_ref, Table)
+            else:
+                self._table = odxlinks.resolve(self.table_ref)
+
+        self._table_row: Optional[TableRow] = None
         if self.table_row_ref:
-            self._table_row = odxlinks.resolve(self.table_row_ref)
+            if TYPE_CHECKING:
+                self._table_row = odxlinks.resolve(self.table_row_ref, TableRow)
+            else:
+                self._table_row = odxlinks.resolve(self.table_row_ref)
 
     def _resolve_snrefs(self, diag_layer: "DiagLayer") -> None:
         super()._resolve_snrefs(diag_layer)
 
         if self.table_snref:
-            self._table = diag_layer.tables[self.table_snref]
+            ddd_spec = diag_layer.diag_data_dictionary_spec
+            self._table = ddd_spec.tables[self.table_snref]
         if self.table_row_snref:
-            self._table_row = diag_layer.tables[self.table_row_snref]
+            # make sure that we know the table to which the table row
+            # SNREF is relative to.
+            assert self._table is not None, "If a table-row short name " \
+                "reference is defined, a table must also be specified."
+            self._table_row = self._table.table_rows[self.table_row_snref]
 
     @property
     def table(self) -> "Table":
         return self._table
 
     @property
-    def table_row(self) -> "TableRow":
+    def table_row(self) -> Optional["TableRow"]:
```

Comment:
Why optional?

Reply:
because table keys with a defined table row is the static case (i.e., in this case, the interpretation of the associated TABLE-STRUCT-PARAMs does not depend on the received message.)